kombu/transport/redis.py

```diff
@@ -829,23 +831,45 @@ def get_table(self, exchange):
                 raise InconsistencyError(NO_ROUTE_ERROR.format(exchange, key))
             return [tuple(bytes_to_str(val).split(self.sep)) for val in values]
 
-    def _lookup_direct(self, exchange, routing_key):
-        if not exchange:
-            return [routing_key]
+    def _lookup(self, exchange, routing_key, default=None):
+        """Find all queues matching `routing_key` for the given `exchange`.
 
-        key = self.keyprefix_queue % exchange
-        pattern = ''
-        queue = routing_key
-        queue_bind = self.sep.join([
-            routing_key or '',
-            pattern,
-            queue or '',
-        ])
-        with self.conn_or_acquire() as client:
-            if client.sismember(key, queue_bind):
-                return [queue]
+        Returns:
+            str: queue name -- must return the string `default`
+                if no queues matched.
+        """
+        result = []
+        if default is None:
+            default = self.deadletter_queue
+        if not exchange:  # anon exchange
+            return [routing_key or default]
+
+        try:
+            pattern = ''
+            queue = routing_key
+            redis_key = self.keyprefix_queue % exchange
+            queue_bind = self.sep.join([
+                routing_key or '',
+                pattern,
+                queue or '',
+            ])
+            with self.conn_or_acquire() as client:
+                if client.sismember(redis_key, queue_bind):
+                    result = [queue]
+                if not client.scard(redis_key):
+                    raise InconsistencyError(NO_ROUTE_ERROR.format(
```

Comment:
Are we sure we want to raise an exception instead of returning `[]` as before?
Why change this?

See https://github.com/celery/kombu/blob/v4.3.0/kombu/transport/virtual/base.py#L718 for example.

Reply:
Yes, it is subtle but that method you referenced calls lookup which has as one of its arguments self.get_table(exchange) -- which will raise this InconsistencyError if there are no members in the set matching that key.   This is why I generally prefer that code not be so deeply nested -- I'd prefer that the call in base define the variable name that is get_table and pass that in, because it becomes self documenting code -- its not clear initially to the reader without much analysis that get_table is going to return all members in the set, nor is it clear that the call to lookup might raise an InconsistencyError but really anything that might raise its own error should probably be on its own line instead of tightly nested in a function call.

Reply:
https://github.com/celery/kombu/blob/master/kombu/transport/redis.py#L829

Reply:
What I am mostly unclear on is if the patrern is actually always going to be an empty string '' for redis?